README.md
Team and repository tags
========================

[![Team and repository tags](http://governance.openstack.org/badges/fuel-astute.svg)](http://governance.openstack.org/reference/tags/index.html)

<!-- Change things from this point on -->

Astute
======

Astute is orchestrator, which is using data about nodes and deployment settings performs two things:
- provision
- deploy

Provision
-----

OS installation on selected nodes.

Provisioning is done using Cobbler. Astute orchestrator collects data about nodes and creates corresponding Cobbler systems using parameters specified in engine section of provision data. After the systems are created, it connects to Cobbler engine and reboots nodes according to the power management parameters of the node.

Deploy
-----

OpenStack installation in the desired configuration on the selected nodes.

Astute uses data about nodes and deployment settings and recalculates parameters needed for deployment. Calculated parameters are passed to the nodes being deployed by use of nailyfact MCollective agent that uploads these attributes to `/etc/astute.yaml` file of the node. Then puppet parses this file using Facter plugin and uploads these facts into puppet. These facts are used during catalog compilation phase by puppet. Finally catalog is executed and Astute orchestrator passes to the next node in deployment sequence. Fuel Library provides puppet modules for Astute.

Using as library
-----

```ruby
require 'astute'

class ConsoleReporter
  def report(msg)
    puts msg.inspect
  end
end

reporter = ConsoleReporter.new

orchestrator = Astute::Orchestrator.new(log_parsing=false)

# Add systems to cobbler, reboot and start installation process.
orchestrator.provision(reporter, environment['engine'], environment['nodes'])

# Observation OS installation
orchestrator.watch_provision_progress(reporter, environment['task_uuid'], environment['nodes'])

# Deploy OpenStack
orchestrator.deploy(reporter, environment['task_uuid'], environment['nodes'])

```
Example of using Astute as library: lib/astute/server/dispatcher.rb


Using as CLI
-----

CLI interface in Astute no longer supported. Please use new Fuel-CLI. More details you can get by link: https://github.com/openstack/fuel-docs/blob/master/pages/user-guide/cli.rst

-----

- ISO, other materials: http://fuel.mirantis.com/
- User guide: http://docs.mirantis.com/
- Development documentation: http://docs.mirantis.com/fuel-dev/


License
------

Licensed under the Apache License, Version 2.0 (the "License"); you may
not use this file except in compliance with the License. You may obtain
a copy of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
License for the specific language governing permissions and limitations
under the License.


debian/changelog
astute (11.0.0-1) unstable; urgency=low

  * Bump version to 11.0.0

 -- Vladimir Kuklin <vkuklin@mirantis.com>  Thu, 02 Mar 2017 15:49:12 +0300

astute (10.0.0-1) trusty; urgency=low

  * Bump version to 10.0

 -- Sergey Kulanov <skulanov@mirantis.com>  Mon, 21 Mar 2016 13:49:10 +0200

astute (9.0.0-1) trusty; urgency=low

  * Update version to 9.0.0

 -- Sergey Kulanov <skulanov@mirantis.com>  Thu, 17 Dec 2015 15:35:14 +0200

astute (8.0.0-1) trusty; urgency=low

  * Update version to 8.0.0

 -- Vladimir Sharshov <vsharshov@mirantis.com>  Fri, 28 Aug 2015 13:30:00 +0300

astute (7.0.0-1) trusty; urgency=low

  * Update version to 7.0.0

 -- Aleksandra Fedorova <afedorova@mirantis.com>  Mon, 08 Jun 2015 19:30:00 +0300

astute (6.1.0-1) trusty; urgency=low

  * Update version to 6.1.0

 -- Matthew Mosesohn <mmosesohn@mirantis.com>  Wed, 22 Apr 2015 14:44:00 +0300

astute (6.0.0-1) trusty; urgency=low

  * Update code from upstream

 -- Igor Kalnitsky <ikalnitsky@mirantis.com>  Wed, 26 Nov 2014 19:49:00 +0200

astute (0.0.1-ubuntu1) precise; urgency=low

  * Update code from upstream

 -- OSCI Jenkins <dburmistrov@mirantis.com>  Wed, 03 Sep 2014 15:20:13 +0400

astute (0.0.1) unstable; urgency=low

  * Initial release.

 -- Mirantis Product <product@mirantis.com>  Tue, 20 Aug 2013 22:20:46 +0400


examples/convert.rb
#!/usr/bin/env ruby

#    Copyright 2013 Mirantis, Inc.
#
#    Licensed under the Apache License, Version 2.0 (the "License"); you may
#    not use this file except in compliance with the License. You may obtain
#    a copy of the License at
#
#         http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
#    License for the specific language governing permissions and limitations
#    under the License.


require 'json'
require 'yaml'

public_iface = "eth0"
internal_iface = "eth1"
private_iface = "eth2"
default_gateway = "172.18.94.33"
master_ip = "172.18.94.34"
dns1 = default_gateway
internal_net = '10.107.2.0'

internal_net_prefix = internal_net.split('.')[0..2].join(".")


nodes = {"compute-01" => '64:c3:54:54:d2:66',
         "controller-01" => "64:48:7a:14:83:e8",
         "controller-02" => "64:b7:37:b1:1d:c9",
         "controller-03" => "64:f4:64:e7:50:d3",
         "swift-01"      => "64:57:26:83:1d:ca",
         "swift-02"      => "64:dc:fd:ad:eb:4e",
         "swift-03"      => "64:ea:df:59:79:39",
         "swiftproxy-01" => "64:bc:c3:9c:07:26",
         "swiftproxy-02" => "64:97:93:5f:b2:dc"
}

template = YAML.load(File.open('example_new.yaml'))
template_node = template['node_01']
newyaml = template
newyaml['nodes'] = []
newyaml.delete('node_01')
cluster = JSON.load(File.open('full.json'))
nodes.each do |node,macaddr| 
    result = template_node.clone
    json_node = cluster.select {|n| n['mac'].to_s == macaddr.to_s.upcase}[0]
    mac = json_node['mac'].to_s
    ip = json_node['ip'].to_s
    l_octet = ip.split('.')[3]
    id = json_node['id'].to_s
    uid = id
    if node == nodes.select{ |n,m| n.to_s =~ /controller/ }.map{|n,m| n}.first
        role = 'primary-controller'
    elsif node =~ /controller/
        role = 'controller'
    elsif node == nodes.select { |n,m| n.to_s =~ /swiftproxy/ }.map{|n,m| n}.first
        role = 'primary-swift-proxy'
    elsif node =~ /swiftproxy/
        role = 'swift-proxy'
    elsif node =~ /swift-\d+/
        role = 'storage'
    else
        role = 'compute'
    end

    cobbler_dnsname = "#{node}.domain.tld"
    cobbler_interfaces = {
        public_iface => {"ip_address"=>ip, "netmask"=> "255.255.255.0", "dns_name"=>cobbler_dnsname, "static"=> "1", "mac_address" => mac}
    }
    cobbler_interfaces_extra = {
        public_iface => {'onboot'=>'yes','peerdns'=>'no'},
        internal_iface => {'onboot'=>'no','peerdns'=>'no'},
        private_iface => {'onboot'=>'no','peerdns'=>'no'}
    }
    result['interfaces'] = cobbler_interfaces
    result['interfaces_extra'] = cobbler_interfaces_extra
    result['power_address'] = ip
    result['mac'] = mac
    result['default_gateway'] = default_gateway
    result['name'] = node
    result['ip'] = ip
    result['id'] = id
    result['uid'] = uid
    result['name_servers'] = master_ip
    result['roles'] = [role]
    result['fqdn'] = cobbler_dnsname
    system_disk=json_node['meta']['disks'].select {|disk| disk['name'] == 'vda'}.first
    cinder_disk=json_node['meta']['disks'].select {|disk| disk['name'] == 'vdb'}.first

    system_disk_path = system_disk['disk']
    system_disk_size = (system_disk['size']/1048756.0).floor
    cinder_disk_path = cinder_disk['disk']
    cinder_disk_size = (cinder_disk['size']/1048756.0).floor

    system_pv_size = system_disk_size - 201
    swap_size = 1024
    free_vg_size = system_pv_size - swap_size
    free_extents = (free_vg_size/32.0).floor
    system_disk_size = 32 * free_extents


#    ks_spaces: '"[{\"type\": \"disk\", \"id\": \"disk/by-path/pci-0000:00:06.0-virtio-pci-virtio3\",
#     \"volumes\": [{\"mount\": \"/boot\", \"type\": \"partition\", \"size\": 200},
#     {\"type\": \"mbr\"}, {\"size\": 20000, \"type\": \"pv\", \"vg\": \"os\"}],
#     \"size\": 20480}, {\"type\": \"vg\", \"id\": \"os\", \"volumes\": [{\"mount\":
#     \"/\", \"type\": \"lv\", \"name\": \"root\", \"size\": 10240 }, {\"mount\":
#     \"swap\", \"type\": \"lv\", \"name\": \"swap\", \"size\": 2048}]}]"'
 

    ks_spaces = '"[{\"type\": \"disk\", \"id\": \"' +
      system_disk_path.to_s +
      '\",\"volumes\": [{\"mount\": \"/boot\", \"type\": \"partition\", \"size\": 200}, {\"type\": \"mbr\"}, {\"size\": ' +
      system_pv_size.to_s +
      ', \"type\": \"pv\", \"vg\": \"os\"}],\"size\": ' +
      system_disk_size.to_s +
      '},{\"type\": \"vg\", \"id\": \"os\", \"volumes\": [{\"mount\": \"/\", \"type\": \"lv\", \"name\": \"root\", \"size\": ' +
      system_disk_size.to_s +
      '},  {\"mount\": \"swap\", \"type\": \"lv\", \"name\": \"swap\", \"size\": '+
      swap_size.to_s +
      '}]}, {\"type\": \"disk\", \"id\": \"' + cinder_disk_path + '\", \"volumes\":  [{\"type\": \"mbr\"}, {\"size\": ' +
      cinder_disk_size.to_s +
      ', \"type\": \"pv\", \"vg\": \"cinder-volumes\"}], \"size\": ' +
      cinder_disk_size.to_s + '}]"'
  
        
    cobbler_ks_meta={"ks_spaces"=>ks_spaces,"mco_host"=>master_ip}

    result['ks_meta'] = result['ks_meta'].update(cobbler_ks_meta)
    puppet_network_data = [
        {"name" => 'public', 'ip'=>ip, "dev" => public_iface, 'netmask' => "255.255.255.0", "gateway" => default_gateway },
        {"name" => ['management','storage'], 'ip'=>"#{internal_net_prefix.to_s}.#{l_octet}", "dev" => internal_iface, 'netmask' => "255.255.255.0"},
        {"name" => 'fixed', "dev" => private_iface},
    ]
    result['network_data'] = puppet_network_data
#    puts result.to_yaml
    newyaml['nodes'].push(result)
end

   newyaml['attributes']['master_ip'] = master_ip
   newyaml['attributes']['dns_nameservers'] = master_ip
   newyaml['attributes']['libvirt_type'] = 'kvm'
   newyaml['attributes']['public_vip'] = '172.18.94.46' 
   newyaml['attributes']['management_vip'] = '10.107.2.254' 
   newyaml['attributes']['floating_network_range'] = '172.18.94.48/28' 
   newyaml['attributes']['fixed_network_range'] = '10.107.2.0/24' 
   newyaml['attributes']['base_syslog']['syslog_server'] = master_ip

puts newyaml.to_yaml

